core/iwasm/libraries/lib-wasi-threads/stress_test/mutex_common.h

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright (C) 2023 Amazon.com Inc. or its affiliates. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+
+#ifndef MUTEX_COMMON_H
+#define MUTEX_COMMON_H
+
+#include <pthread.h>
+#include <stdio.h>
+#include <assert.h>
+#include <errno.h>
+#include <unistd.h>
+#include <stdbool.h>
+
+typedef struct {
+    pthread_mutex_t *mutex;
+    int *counter;
+    int limit;
+    bool is_sleeping;
+} MutexCounter;
+
+enum constants {
+    NUM_ITER = 100000,
+    NUM_THREADS = 12,
+    NUM_RETRY = 8,
+    RETRY_SLEEP_TIME_US = 1000,
+};
+
+// This enum defines whether thread should sleep to increase contention
+enum sleep_state {
+    NON_SLEEP = 0,
+    SLEEP = 1,
+};
+
+void
+spawn_thread(pthread_t *tid, void *func, void *arg)
+{
+    int status_code = -1;
+    int timeout_us = RETRY_SLEEP_TIME_US;
+    for (int tries = 0; status_code != 0 && tries < NUM_RETRY; ++tries) {
+        status_code = pthread_create(tid, NULL, (void *(*)(void *))func, arg);
+        assert(status_code == 0 || status_code == EAGAIN);
+        if (status_code == EAGAIN) {
+            usleep(timeout_us);
+            timeout_us *= 2;
+        }
+    }
+
+    assert(status_code == 0 && "Thread creation should succeed");
+}
+
+// We're counting how many times each thread was called using this array
+// Main thread is also counted here so we need to make arrays bigger
+typedef struct {
+    int tids[NUM_THREADS + 1];
+    int calls[NUM_THREADS + 1];
+} StatCollector;
+
+StatCollector call_stat;
+
+void
+clear_stat()
+{
+    for (int i = 0; i < NUM_THREADS + 1; ++i) {
+        call_stat.tids[i] = 0;
+        call_stat.calls[i] = 0;
+    }
+}
+
+void
+add_to_stat(int tid)
+{
+    int tid_num = 0;
+    for (; tid_num < NUM_THREADS + 1 && call_stat.tids[tid_num] != 0;
+         ++tid_num) {
+        if (call_stat.tids[tid_num] == tid) {
+            call_stat.calls[tid_num]++;
+            return;
+        }
+    }
+
+    assert(tid_num < NUM_THREADS + 1);
+    call_stat.tids[tid_num] = tid;
+    call_stat.calls[tid_num] = 1;
+}
+
+// This function ensures that every of NUM_THREADS+1 threads number of
+// operations is in between [NUM_ITER/NUM_THREADS, NUM_ITER/(NUM_THREADS+2)].
+// For 100'000 operations it means that every thread should be called [7142,
```

Comment:
why are we checking that?

Reply:
why not to? 

I mean if you have more or less random sleeps and still spending 50% of operations in one TID probably smth is bad with your system

Reply:
That **might** be true if you run this test a number of times - than **probably** in average that's what you will achieve. However, keeping it like that is asking for flakiness. If you  want to test this behavior, and do it based on statistics, I'd rather move it to separate test suite which is expected to sometimes fail (but then we'd need to figure out how to track this test suite).

We can keep it like that, but as I said - expect flakiness here

Reply:
I didn't see any flakiness here running the tests locally and in CI  but I trust your experience so I'll remove this check : )